tests/ui/wrong_self_convention2.rs

```diff
@@ -87,3 +87,12 @@ mod issue3414 {
         }
     }
 }
+
+mod issue4546 {
+    use std::pin::Pin;
+
+    struct S;
+    impl S {
+        pub fn as_mut(self: Pin<&mut Self>) {}
```

Comment:
It would be nice to also have a test for `Pin<&Self>` and some test for the other listed cases from the [lint documentation](https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention). (I haven't tested these, so there might still be some false positives with `to_`)

```suggestion
        pub fn as_mut(self: Pin<&mut Self>) {}

        pub fn as_other_thingy(self: Pin<&Self>) {}

        pub fn to_mut(self: Pin<&mut Self>) {}

        pub fn to_other_thingy(self: Pin<&Self>) {}
```